crates/core/src/transport/peer_connection.rs

```diff
@@ -351,6 +351,28 @@ impl PeerConnection {
         loop {
             // tracing::trace!(remote = ?self.remote_conn.remote_addr, "waiting for inbound messages");
             tokio::select! {
+                // Check completed streams first to prevent channel backup
+                inbound_stream = self.inbound_stream_futures.next(), if !self.inbound_stream_futures.is_empty() => {
```

Comment:
select does not work like that, moving the branch up and down does not change prioritazion, it still is random: 
https://docs.rs/tokio/latest/tokio/macro.select.html#fairness


you need to use the biassed keyword. Hoever, IF we do this change really monitor this closely cause inadvertidly we may break other part of the system